### PR TITLE
Drop `versionize` support, substitute with `serde`

### PR DESCRIPTION
**Pull request by bchalios**
## Changes

Remove support for `Versionize` from Firecracker and substitute it with `serde`. We use `bincode` for the actual (de)serialization. Also, define a snapshot format version that is decoupled from Firecracker version. The newly defined version of the snapshot format will follow semantic versioning and it will start from version `1.0.0`.

This PR breaks compatibility with all previous snapshot versions. From now on, snapshot compatibility will be reflected to the snapshot version. Changes in the major release number are backwards incompatible, changes in the minor release number are backwards compatible and changes in patch releases are always compatible.

That said, we are currently using `bincode` for the serialization, which means that all changes in the snapshot format will be backwards incompatible (they will imply a major version number bump). However, if the need arises we will investigate alternatives to `bincode` that would allow greater flexibility for backwards compatibility.

This change makes a lot of our integration tests redundant.

## Reason

We used `Versionize` to allow for extended backwards compatibility for Firecracker snapshots. In practice we did not find this feature useful, so we paid the code maintenance and testing complexity cost without a measurable benefit.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license. For more information on following
Developer Certificate of Origin and signing off your commits, please check
[`CONTRIBUTING.md`][3].

## PR Checklist

- [x] If a specific issue led to this PR, this PR closes the issue.
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] API changes follow the [Runbook for Firecracker API changes][2].
- [x] User-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.
- [x] New `TODO`s link to an issue.
- [x] Commits meet [contribution quality standards](https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md#contribution-quality-standards).

---

- [ ] This functionality cannot be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md
[3]: https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md

### CODE REVIEW

**Comment by bchalios** [comment]
This PR depends on the equivalent PR on kvm-bindings side: https://github.com/firecracker-microvm/kvm-bindings/pull/21

**Comment by bchalios** [comment]
I deliberately split up the commits that switch types from `Versionize` to `serde` so that I facilitate the review. but I am happy to merge these in a single commit if we decide its cleaner that way.

File: docs/snapshotting/versioning.md
Comment (roypat) [inline docs/snapshotting/versioning.md:60]: Wait, is this really semantic versioning? Shouldn't a Firecracker binary that declares support for snapshot version format, say, 2.1.0 also be able to load version, say, 2.5.0?

Also, this should say something like "will be able to load all snapshot files with version `x.Y.Z` with `Y<y` or `Y=y and Z<=z`.

File: src/vmm/src/arch/aarch64/regs.rs
Comment (ShadowCurse) [reply src/vmm/src/arch/aarch64/regs.rs:429]: It is used to restore old snap versions. Considering we are removing versionize, we can remove this type + remove `old_regs` field from `VcpuState` for aarch64.

File: src/vmm/src/arch/aarch64/gic/regs.rs
Comment (bchalios) [reply src/vmm/src/arch/aarch64/gic/regs.rs:55]: silly question probably, but is it passing on `aarch64`?

File: src/vmm/src/arch/aarch64/regs.rs
Comment (bchalios) [reply src/vmm/src/arch/aarch64/regs.rs:318]: I had an offline discussion with @ShadowCurse about this. Apparently, the check for <= 2048 bits made sense only for cases where we would have snapshots from a different version than the one we are loading the snapshot for. But I realize now that the check for the total size of the registers might be important (?)

File: src/vmm/src/arch/aarch64/gic/regs.rs
Comment (roypat) [reply src/vmm/src/arch/aarch64/gic/regs.rs:55]: Yeah, `cargo check --aarch64-unknown-linux-gnu` passes. We also dont implement serialize and deserialize for any of the types that implement the trait. I think it works because we dont ever use this trait in a generic fashion during (de)serialization, and all concrete types for which it is implemented provide a `RegChunk` type that is serializable.

File: src/vmm/src/arch/aarch64/regs.rs
Comment (ShadowCurse) [reply src/vmm/src/arch/aarch64/regs.rs:318]: Technically this check is to prevent snapshot tampering, but we already have crc check for it, so I think we can remove it. Also snapshots are generally trusted by us.

Comment (roypat) [reply src/vmm/src/arch/aarch64/regs.rs:318]: I remember that we needed this check to get Firecracker to gracefully exit instead of exit with panic if an invalid snapshot file is passed. If that's no longer the case, then we can remove it, but if without this, Firecracker can be made to panic, we should keep it

File: src/snapshot/src/lib.rs
Comment (ShadowCurse) [inline src/snapshot/src/lib.rs:344]: Shouldn't this be `1.4.12` that is not supported because it is bigger than `1.3.12`?

File: src/snapshot/src/lib.rs
Comment (ShadowCurse) [inline src/snapshot/src/lib.rs:357]: And here it should be `1.2.12` as supported as it is less than `1.3.12`

File: Cargo.lock
Comment (bchalios) [reply Cargo.lock]: Yup, I think this was due to the fact that we kept using `Vesionize` which was bringing its own version of the crate. I didn't remove it at the time.

File: src/vmm/src/arch/aarch64/regs.rs
Comment (bchalios) [reply src/vmm/src/arch/aarch64/regs.rs:318]: Under which condition would this cause Firecracker to panic? Right now, we only snapshot and restore from the same Firecracker version. Under this condition, I don't see how this could fail. Am I missing something?

Comment (ShadowCurse) [reply src/vmm/src/arch/aarch64/regs.rs:318]: Oh yeah, I think I remember now. It is because of this [`unreachable`](https://github.com/firecracker-microvm/firecracker/blob/bccf2d4ebaa54265584fb23898fdb43cf92bd75c/src/vmm/src/arch/aarch64/regs.rs#L171 ) If snapshot was modified and the id of some register is changed to have values bigger than 8 in the reg size position then conversion will fail. (8 because 2^8 = 256 bytes -> 2048 bit registers - the max size kernel supports)

Comment (bchalios) [reply src/vmm/src/arch/aarch64/regs.rs:318]: Ok, I will just re-introduce the manual deserialization. It's not that much of a trouble and we will avoid any surprises in the future.

File: src/vmm/src/arch/aarch64/gic/regs.rs
Comment (bchalios) [reply src/vmm/src/arch/aarch64/gic/regs.rs]: done!

File: src/vmm/src/arch/aarch64/regs.rs
Comment (bchalios) [reply src/vmm/src/arch/aarch64/regs.rs]: done

File: src/vmm/src/arch/aarch64/regs.rs
Comment (bchalios) [reply src/vmm/src/arch/aarch64/regs.rs:318]: Ok, manual implementations are not as straight-forward as I thought. I will look into it again after I address the rest of the comments.

File: src/vmm/src/arch/aarch64/regs.rs
Comment (bchalios) [reply src/vmm/src/arch/aarch64/regs.rs]: done

File: tests/integration_tests/functional/test_cmd_line_parameters.py
Comment (bchalios) [reply tests/integration_tests/functional/test_cmd_line_parameters.py:42]: Problem is it can't. The snapshot header has changed in a way that it doesn't allow it parse older snapshot versions.

But you are right. I think the correct thing to do here is to skip this for older supported releases. WDYT?

File: tests/integration_tests/functional/test_snapshot_basic.py
Comment (bchalios) [reply tests/integration_tests/functional/test_snapshot_basic.py:57]: It seems that `run_cmd_sync` does two different things depending on whether the command we pass is a list or not, although I'm not sure what's the difference between the two.

Comment (bchalios) [reply tests/integration_tests/functional/test_snapshot_basic.py:57]: In any case, I did the change :P

File: tests/integration_tests/performance/test_snapshot_perf.py
Comment (bchalios) [reply tests/integration_tests/performance/test_snapshot_perf.py:1]: true! Are there not unused code warnings for Python?

File: tests/integration_tests/functional/test_snapshot_advanced.py
Comment (bchalios) [reply tests/integration_tests/functional/test_snapshot_advanced.py:1]: true! Hadn't noticed. Removed.

File: tests/integration_tests/functional/test_cmd_line_parameters.py
Comment (roypat) [reply tests/integration_tests/functional/test_cmd_line_parameters.py]: > I think the correct thing to do here is to skip this for older supported releases.

Yeah, this is what I meant, sorry :)

File: tests/integration_tests/performance/test_snapshot_perf.py
Comment (roypat) [reply tests/integration_tests/performance/test_snapshot_perf.py:1]: Sadly no :(
I suppose its because you can technically do something absurd such as constructing a function name via string concatenation, then call it via looking it up in `__dict__` or `globals()`. So you can only find out which functions are called by actually running the code. 
That's why we had PRs in the past that deleted like 500 lines of unused python code :joy:

File: tests/integration_tests/functional/test_cmd_line_parameters.py
Comment (bchalios) [reply tests/integration_tests/functional/test_cmd_line_parameters.py]: Done :)

Comment (roypat) [reply tests/integration_tests/functional/test_cmd_line_parameters.py]: Shouldn't the diff to the test be?

```diff
diff --git a/tests/integration_tests/functional/test_cmd_line_parameters.py b/tests/integration_tests/functional/test_cmd_line_parameters.py
index 59e26e4e9..99d17ca03 100644
--- a/tests/integration_tests/functional/test_cmd_line_parameters.py
+++ b/tests/integration_tests/functional/test_cmd_line_parameters.py
@@ -7,26 +7,22 @@ from pathlib import Path
 
 import pytest
 
-from framework.utils import run_cmd
+from framework.utils import run_cmd, get_firecracker_version_from_toml
 from host_tools.cargo_build import get_firecracker_binaries
 from host_tools.metrics import validate_fc_metrics
 
 
-def test_describe_snapshot_all_versions(
-    microvm_factory, guest_kernel, rootfs, firecracker_release
-):
+def test_describe_snapshot_all_versions(microvm_factory, guest_kernel, rootfs):
     """
     Test `--describe-snapshot` correctness for all snapshot versions.
 
     For each release create a snapshot and verify the data version of the
     snapshot state file.
     """
-    target_version = firecracker_release.snapshot_version
+    target_version = get_firecracker_version_from_toml()
     vm = microvm_factory.build(
         guest_kernel,
         rootfs,
-        fc_binary_path=firecracker_release.path,
-        jailer_binary_path=firecracker_release.jailer,
     )
     vm.spawn()
     vm.basic_config(track_dirty_pages=True)

```

Comment (bchalios) [reply tests/integration_tests/functional/test_cmd_line_parameters.py]: I _think_ not. IIUC the point of this test, it is to check if `firecracker --describe-snapshot` works correctly for all previously supported snapshot formats. Once we start having more Firecracker binaries that "speak" the new snapshot format this test will create a snapshot for every previously supported Firecracker version (>= 1.6.0) and then use the latest `main` to check that `--describe-snapshot` works for all of the previous snapshots.

File: docs/snapshotting/snapshot-support.md
Comment (bchalios) [reply docs/snapshotting/snapshot-support.md]: done!

File: tests/integration_tests/functional/test_cmd_line_parameters.py
Comment (roypat) [reply tests/integration_tests/functional/test_cmd_line_parameters.py]: Ahhhh, I get it now. Didnt catch that it used the current firecracker version to call `--describe-snapshot`. Okay discard what I said :D

File: src/snapshot/src/lib.rs
Comment (bchalios) [reply src/snapshot/src/lib.rs:344]: `1.2.12` here represents the snapshot version that a Firecracker binary supports. `1.3.12` is the version of the saved snapshot.

File: src/snapshot/src/lib.rs
Comment (bchalios) [reply src/snapshot/src/lib.rs:357]: See last comment

File: src/snapshot/src/lib.rs
Comment (bchalios) [reply src/snapshot/src/lib.rs:77]: done

File: src/snapshot/src/lib.rs
Comment (ShadowCurse) [reply src/snapshot/src/lib.rs:344]: ok, I get it now.

File: src/vmm/src/arch/aarch64/regs.rs
Comment (ShadowCurse) [reply src/vmm/src/arch/aarch64/regs.rs:318]: Maybe you can leave the wrapper type for now and we create a `good new issue` to remove it?

Comment (bchalios) [reply src/vmm/src/arch/aarch64/regs.rs]: Ok, done! I combined your idea Egor with an idea of Patrick's where I just serialize manually a tuple including the fields and then deserialize that tuple and perform the checks.

File: docs/snapshotting/versioning.md
Comment (bchalios) [reply docs/snapshotting/versioning.md]: I removed mentions of semantic versioning. In any case, with bincode we will not be able to make any backwards compatible changes to the snapshot format. If in the future we do, we should define this properly, for the moment we just mention the version format `MAJOR.MINOR.PATCH` which is generic enough to describe whatever backwards compatibility we need.

For the record, SchemaVer (which is based on semantic versioning) seems to fit well to what we have in mind: https://snowplow.io/blog/introducing-schemaver-for-semantic-versioning-of-schemas/

File: tests/framework/artifacts.py
Comment (bchalios) [reply tests/framework/artifacts.py:109]: I added a TODO in both places where I do these checks. I will open the Github issues to remove these, once 1.5 goes out of support.

File: docs/snapshotting/versioning.md
Comment (roypat) [reply docs/snapshotting/versioning.md]: Sounds good!

**Comment by roypat** [comment]
I think https://github.com/firecracker-microvm/firecracker/blob/da92bf6d4afa678565dcf9df2e4e00d2fa273d96/src/vmm/src/persist.rs#L162-L165 can also be deleted :)

File: src/vmm/src/arch/aarch64/regs.rs
Comment (bchalios) [reply src/vmm/src/arch/aarch64/regs.rs]: done!

File: Cargo.toml
Comment (bchalios) [reply Cargo.toml:11]: done!

File: src/snapshot/src/lib.rs
Comment (bchalios) [reply src/snapshot/src/lib.rs]: fixed this instance as well.

File: src/vmm/src/devices/virtio/persist.rs
Comment (ShadowCurse) [reply src/vmm/src/devices/virtio/persist.rs:339]: Maybe create an issue to remove these later to limit changes in this PR?

**Comment by roypat** [comment]
@ShadowCurse do you know if the "needed for compatibility" enum variants in https://github.com/firecracker-microvm/firecracker/blob/main/src/vmm/src/cpu_config/aarch64/static_cpu_templates/mod.rs#L14-L19 can be removed now that versionize is going away?

File: src/vmm/src/vmm_config/boot_source.rs
Comment (roypat) [inline src/vmm/src/vmm_config/boot_source.rs:35]: mh, why remove this? Or rather, why was this here in the first place?

Actually, why did these already implement Serialize anyway. They're for the API, so only Deserialize should've been needed :thinking:

File: src/vmm/src/vstate/memory.rs
Comment (roypat) [inline src/vmm/src/vstate/memory.rs:575]: Mh, do we even want to serialize the dirty bitmap? After a snapshot is taken it should get reset to "everything is clean" (because we just took a snapshot)

**Comment by roypat** [comment]
You missed a handful of the "requires version bump" comments, I think:

```
$ grep "version bump" src -R
src/vmm/src/arch/aarch64/gic/gicv2/regs/dist_regs.rs:// are not saved. NOTICE: Any changes to this structure require a snapshot version bump.
src/vmm/src/arch/aarch64/gic/gicv2/regs/icc_regs.rs:// NOTICE: Any changes to this structure require a snapshot version bump.
src/vmm/src/arch/aarch64/gic/gicv3/regs/redist_regs.rs:// NOTICE: Any changes to this structure require a snapshot version bump.
src/vmm/src/arch/aarch64/gic/gicv3/regs/redist_regs.rs:// NOTICE: Any changes to this structure require a snapshot version bump.
src/vmm/src/arch/aarch64/gic/gicv3/regs/dist_regs.rs:// snapshot version bump.
src/vmm/src/arch/aarch64/gic/gicv3/regs/icc_regs.rs:// NOTICE: Any changes to this structure require a snapshot version bump.
src/vmm/src/arch/aarch64/gic/gicv3/regs/icc_regs.rs:// NOTICE: Any changes to this structure require a snapshot version bump.
```

File: src/vmm/src/device_manager/persist.rs
Comment (roypat) [inline src/vmm/src/device_manager/persist.rs:171]: maybe `V2` is a more sensible default?

File: tests/integration_tests/functional/test_api.py
Comment (roypat) [inline tests/integration_tests/functional/test_api.py:1196]: ah, this answers my first question :)

I guess this is fine because we are adding a new field to an existing response, which can be done in a minor

**Comment by ShadowCurse** [comment]
> @ShadowCurse do you know if the "needed for compatibility" enum variants in https://github.com/firecracker-microvm/firecracker/blob/main/src/vmm/src/cpu_config/aarch64/static_cpu_templates/mod.rs#L14-L19 can be removed now that versionize is going away?

Hmm... actually by removing `verisonize` we break compatibility, so I think this is a good time to fix `StaticCpuTemplate` enum (for both x86 and aarch64). The reason the enum looks so strange is the historical reason that goes back to the time when we did not have aarch64 templates, but still were storing `StaticCpuTemplate` in the aarch64 snapshots.

I think we should clean `StaticCpuTemplate` from `Empty#` fields and `None` case maybe. But this can be done in a separate PR (it is not really a `versionize` problem). We just need not to forget.

**Comment by roypat** [comment]
> > @ShadowCurse do you know if the "needed for compatibility" enum variants in https://github.com/firecracker-microvm/firecracker/blob/main/src/vmm/src/cpu_config/aarch64/static_cpu_templates/mod.rs#L14-L19 can be removed now that versionize is going away?
> 
> Hmm... actually by removing `verisonize` we break compatibility, so I think this is a good time to fix `StaticCpuTemplate` enum (for both x86 and aarch64). The reason the enum looks so strange is the historical reason that goes back to the time when we did not have aarch64 templates, but still were storing `StaticCpuTemplate` in the aarch64 snapshots.
> 
> I think we should clean `StaticCpuTemplate` from `Empty#` fields and `None` case maybe. But this can be done in a separate PR (it is not really a `versionize` problem). We just need not to forget.

I actually already have a local branch that does both, I just need to publish it once this PR is merged xP

File: src/vmm/src/vstate/memory.rs
Comment (bchalios) [reply src/vmm/src/vstate/memory.rs:575]: You are right. We don't want to, and in fact we don't do it. This check does absolutely nothing :) I will remove

**Review by roypat — DISMISSED** [review]

File: src/vmm/src/device_manager/persist.rs
Comment (bchalios) [reply src/vmm/src/device_manager/persist.rs]: In fact, we should completely remove the `Default` implementation here. The state will always be stored in the snapshot file. It doesn't make sense to have this at all, since we will never trigger it.

**Comment by bchalios** [comment]
Changes from v1 -> v2:

* Removed comments I had missed about bumping Versionize version if structs changed
* Removed unnecessary check for serialization/deserialization of GuestMemory dirty tracking bitmap
* Removed default implementation for `MmdsVersionState`. We will always get a valid value from the snapshot so the default implementation was dead code

**Review by roypat — APPROVED** [review]

File: src/vmm/src/persist.rs
Comment (roypat) [inline src/vmm/src/persist.rs:47]: I think this constant can go away :thinking:

Comment (bchalios) [reply src/vmm/src/persist.rs:47]: yup!

**Comment by bchalios** [comment]
Changes v3 -> v4:

* Removed `FC_V0_23_MAX_DEVICES` which was not used any more and an associated Error variant that was only there to express that we were asked to create more than `FC_V0_23_MAX_DEVICES` devices.

**Review by roypat — APPROVED** [review]

File: tests/integration_tests/functional/test_snapshot_basic.py
Comment (pb8o) [inline tests/integration_tests/functional/test_snapshot_basic.py:58]: nit: this could be a method of Microvm class. But since it's not called often it may not be worth it.

File: src/snapshot/benches/version_map.rs
Comment (pb8o) [inline src/snapshot/benches/version_map.rs:1]: Oh removing this will need a change in https://github.com/firecracker-microvm/firecracker/blob/main/tools/bump-version.sh to stop looking for versions in this file. I think it's the only one.

File: src/snapshot/src/crc.rs
Comment (bchalios) [reply src/snapshot/src/crc.rs:1]: What year do we typically put in there? Is it the first time this was written, or the last time it was touched?

File: src/snapshot/benches/version_map.rs
Comment (bchalios) [reply src/snapshot/benches/version_map.rs:1]: Very nice catch. Can you think of another check to replace this with, or should we just drop the check all together?

Comment (bchalios) [reply src/snapshot/benches/version_map.rs:1]: Also, is there a way to test the release scripts?

File: src/snapshot/src/crc.rs
Comment (bchalios) [reply src/snapshot/src/crc.rs:1]: Answering my own question: this should be the year that we first published this code. This was taken from the versionize crate: https://github.com/firecracker-microvm/versionize/blob/main/src/crc.rs, so I think that I should just leave 2020. WDYT?

File: src/snapshot/benches/version_map.rs
Comment (roypat) [reply src/snapshot/benches/version_map.rs:1]: We have a "release sanity build" step in our CI, but it only triggers if specific files in `tools` are touched. You could add a dummy commit that touches `tools/devtool` to trigger it

Comment (pb8o) [reply src/snapshot/benches/version_map.rs:1]: I would go with dropping the check altogether. The issue that this was checking for is gone with your PR, so we don't need it anymore. You can test the release scripts locally by `devtool build --release`, but if you drop the check it will get done in the PR anyway.

File: src/snapshot/src/crc.rs
Comment (pb8o) [reply src/snapshot/src/crc.rs:1]: Ah that's fine by me... I think I have also seen 2020-2024, but it would be the only one in the repo to have a year range.

Comment (bchalios) [reply src/snapshot/src/crc.rs]: Ok, I stayed with the 2020 year. Another issue going the year range is that it would imply that we need to update the range in all the files once per year.

File: src/snapshot/benches/version_map.rs
Comment (bchalios) [reply src/snapshot/benches/version_map.rs:1]: Ok, done

Comment (bchalios) [reply src/snapshot/benches/version_map.rs:1]: it might be worth to run the check in every PR (?)

**Comment by bchalios** [comment]
Changes v4 -> v5:

* Dropped check for versions existing in `src/vmm/src/version_map.rs`; the file doesn't exist any more.
* Added correct year (2020) in copyright notice of `src/vmm/src/snapshot/src/crc.rs` file. This file was taken from the `versionize` crate implementation and 2020 was the first year of publication.

**Review by pb8o — APPROVED** [review]

**Review by ShadowCurse — APPROVED** [review]

**Pull request merged**
